deepspeed/runtime/zero/partitioned_param_coordinator.py

```diff
@@ -430,6 +430,7 @@ def release_and_reset_all(self, module: Module) -> None:
             # there's a hook execution issue
             param.ds_active_sub_modules.clear()
             self.__release_param(param)
+        self.__n_available_params = 0
```

Comment:
This looks like a hack to me. I think it would better to understand why the following decrement is insufficient
https://github.com/microsoft/DeepSpeed/blob/12189feb4f0377dbdbd9337ec14a590b58714f23/deepspeed/runtime/zero/partitioned_param_coordinator.py#L500

Reply:
I explained in detail in my commit log. I copy it here:
"__n_available_params" is set to 0 in reset_step() which is called in
backward(). All parameter are released in release_and_reset_all() which
is called in step(). "__n_available_params" is reduced when parameter is
released. These mean if step() is called after backward(),
"__n_available_params" will be reduced to a negative number.
"__n_available_params" is used to restrict fetched parameters, so
negative value leads to a problem that fetched parameter will be larger
than upper bound ("__max_n_available_params").

Reply:
Generally speaking, "__n_available_params" is set to 0 in reset_step() and this leads to wrong value for following release operation.

Reply:
Thanks for your explanation. Yes, setting `_n_available_params` to 0 in `reset_step()` is a bug since it is inconsistent with the fetched parameter count. Strictly speaking, `_n_ available_params` should be modified by gather (increment) and release (decrement) operations. Since this a pre-existing bug, it is not a requirement that you fix. But it would be greatly appreciated if you could do so. Thanks!

Reply:
Ok, I will update the PR later with unit test together.

Reply:
I found there is a reason to set `__n_available_params=0`. This variable becomes negative somehow after second step. I need to investigative.

Reply:
There is persistent_parameters, which are gathered outside of Class "PartitionedParameterCoordinator". "__n_availiable_params" cannot recode these information, so it becomes negative after release all parameters. That is why there is a line of code to set "__n_availiable_params" to 0. If that is the case, then we need to move "`__n_availiable_params = 0`" to release_and _reset_all() instead of reset_step()